host_core/lib/host_core/actors/actor_module.ex

```diff
@@ -641,39 +653,35 @@ defmodule HostCore.Actors.ActorModule do
     # we consider a hash of bytes as a unique key
     key = :sha256 |> :crypto.hash(bytes) |> Base.encode16()
 
-    module =
-      case :ets.lookup(:module_cache, key) do
-        [{_, cached_mod}] ->
-          cached_mod
-
-        [] ->
-          {:ok, mod} = Wasmex.Module.compile(bytes)
-          :ets.insert(:module_cache, {key, mod})
-          mod
-      end
-
     # TODO - in the future, poll these so we can forward the err/out pipes
     # to our logger
-    {:ok, stdin} = Wasmex.Pipe.create()
-    {:ok, stdout} = Wasmex.Pipe.create()
-    {:ok, stderr} = Wasmex.Pipe.create()
+    {:ok, stdin} = Wasmex.Pipe.new()
+    {:ok, stdout} = Wasmex.Pipe.new()
+    {:ok, stderr} = Wasmex.Pipe.new()
 
-    wasi = %{
+    wasi_opts = %Wasmex.Wasi.WasiOptions{
       args: [],
       env: %{},
-      preopen: %{},
       stdin: stdin,
       stdout: stdout,
       stderr: stderr
     }
 
-    opts =
-      if imports_wasi?(Wasmex.Module.imports(module)) do
-        %{module: module, imports: imports, wasi: wasi}
-      else
-        %{module: module, imports: imports}
+    context =
+      case :ets.lookup(:module_cache, key) do
+        [{_, cached_mod}] ->
+          cached_mod
+
+        [] ->
+          {:ok, store} = Wasmex.Store.new_wasi(wasi_opts)
+          {:ok, mod} = Wasmex.Module.compile(store, bytes)
+          ctx = %{store: store, module: mod}
```

Comment:
Not sure I follow this: does this mean that if I start a second instance of an actor, they'll share the same WASI store/pipes?

Reply:
The store is always given it's own Wasmex.Pipe's. A second instance should mean its own pipes. Once I can actually build and sign WASI actors... I'll add some tests around this. Right now the pipes aren't used and this shouldn't be a change in behavior from our current host.

Reply:
I might be missing something, but the way I read this, if the cache has a matching entry `(store, module)`, we use _both_ the cached store and module. And the store is just one set of pipes, right? I think we want to cache the bytes, but not the store/module now(?)

Reply:
Hope it's OK if I chime in here: A store is more than a bunch of pipes. It internally holds the physical memory for all WASM structures. So `Wasmex.Memory`, `Wasmex.Instance`, and all others are fancy pointers to structures within the Store.

It's a good idea to cache the `store` and `module` together because we need both, the datastore and the pointer to a specific memorry datum in that store, to do wasmex function calls. :)

[The Wasmex docs](https://hexdocs.pm/wasmex/Wasmex.Store.html) as well as the [wasmtime docs](https://github.com/bytecodealliance/wasmtime/blob/d5b8da6eeafc3c99d0404ca764349ec691c72062/crates/wasmtime/src/store.rs#L1) may have more pointers if you want to get into the details.

Reply:
Sorry for the confusion. When I said "just one" I meant it included "one set" of pipes, not that the store was _only_ composed of pipes :)

My concern here is if a user tries to start a second instance of an actor: if we match on the cache key, we will reuse the `module` and `store` from the _previous_ instance of the actor, including its pipes. I don't think that's correct: shouldn't each instance of an actor have its own set of pipes and memory? 

Today I don't think this is an issue, since wasmCloud doesn't use the module's pipes to communicate. But we're considering doing that in the future.

Reply:
> When I said "just one" I meant it included "one set" of pipes, not that the store was only composed of pipes :)

ohhh, yes. sorry i misunderstood you  💡 I'm not too sure about wasmClouds architecture/nomenclature. I understand that one actor is one WASM binary, right? One actor/WASM binary could run many times in parallel.

In theory you could compile and instantiate the WASM binaries in the same `Store` meaning they'd share their IO pipes. But that doesn't sound good as instances would "fight" over these pipes. So, @connorsmith256 , you're right :) Each actor should be in their own store with their own IO pipes. Meaning, they should also have their own Memory struct (can't share them between stores) - so no re-using modules either.

One optimization could be to use `Wasmex.Engine.precompile_module()` to compile a module only once and hydrate it for every new actor (in a fresh store).

Reply:
> One optimization could be to use Wasmex.Engine.precompile_module() to compile a module only once and hydrate it for every new actor (in a fresh store).

This sounds exactly like what we want, a precompiled module that we cache the reference to and instantiate instances from.